Added php7 template (as default base version)

andrew-s opened the pull request:
Signed-off-by: Andrew Seymour <commit@andrewseymour.co.uk>

PHP Template

## Description
Added in PHP7 template (named as PHP - perhaps future versions can be versioned, following the python scheme)

## Motivation and Context
Fixes the request for a PHP template in the default library
- [X] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## Which issue(s) this PR fixes 
Fixes #783

## How Has This Been Tested?
Used extensively in a development environment

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

ivanayov commented on template/php/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+RUN chmod +x /usr/bin/fwatchdog
+
+# Core
+WORKDIR /root/
```
Can you please remove the slash at the end to follow the path conventions?

ivanayov commented on template/php/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+COPY index.php php-extension.sh composer.* ./
+RUN chmod u+x ./php-extension.sh && ./php-extension.sh
+COPY ./function/ ./function/
+WORKDIR /root/function/
```
Here as well

ivanayov commented:
Thanks @andrew-s

Brunty commented on template/php/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+# Composer requirements
+ARG COMPOSER_AUTH='{}'
+ENV COMPOSER_AUTH=${COMPOSER_AUTH}
+RUN curl -sS https://getcomposer.org/installer | php -- --install-dir=/usr/local/bin --filename=composer
```
Would it be worth using multi-stage builds to pull this in from the composer image?

https://medium.com/@othillo/adding-composer-to-php-docker-images-using-multi-stage-builds-2a10967ae6c1

Brunty commented on template/php/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+FROM php:7-alpine
```
Is PHP7 being stuck with?  7.0 is EOL for bug fixes and into security only fixes now.

http://php.net/supported-versions.php

andrew-s replied:
7-alpine is the latest in the 7 branch, I can lock this to say 7.2 if that would be more appropriate otherwise it'll be the latest stable

Brunty replied:
It shouldn't be a problem using the latest, but as PHP doesn't follow semver (can't find sources, have been informed of this in the past though), there could be issues if it just pulls the latest thing when the next `.X` release is out?

Brunty commented on template/php/function/composer.json:
```diff
@@ -0,0 +1,16 @@
+    },
+    "autoload": {
+        "psr-4": {
+            "App\\": "src/"
```
I'd say probably not, the function that'll be written is not an OpenFaaS project (while the templates are) - so it makes more sense to be `App` or `App\OpenFaaS` to denote that it's within the app here, not the OpenFaaS vendor.

affix replied:
Agreed

alexellis replied:
The root / entrypoint is "openfaas" / "com.openfaas" elsewhere. This is overriden in the actual generated template.

andrew-s commented:
Updated the dockerfile with the changes mentioned, thanks for the feedback, updated the watchdog while I was there.

The only thing left to do here, is run composer as non-root but, that's going to take a re-order of that process I think.

alexellis commented on template/php/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+ENV COMPOSER_AUTH=${COMPOSER_AUTH}
+
+# Add git
+RUN apk add --no-cache git
```
I'd like to see all of those instructions combined with `&&` for the benefit of caching.

johnmccabe replied:
This would be best moved into the `RUN` step where the `app` user and homedir are created.

alexellis commented on template/php/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+RUN addgroup -S app && adduser -S -g app app
+RUN mkdir -p /home/app
+
+# Core
```
No need for this comment

alexellis commented on template/php/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+USER app
+WORKDIR /home/app
+ENV fprocess="php index.php"
+HEALTHCHECK --interval=1s CMD [ -e /tmp/.lock ] || exit 1
```
Let's have a 3s check.

alexellis commented on template/php/function/src/Handler.php:
```diff
@@ -0,0 +1,18 @@
+     * @param string $data
+     * @return string
+     */
+    public function handle(string $data): string {
```
Is a string going to be suitable for everyone?

Brunty replied:
It'd be nice to be an array, or abstracted to an object where the request body, as well as any headers etc could be accessed easily.

andrew-s replied:
This is a good point, this would prevent binary data being returned (or accepted). This requirements could be removed and be left to the application to handle.

alexellis replied:
Can you provide an alternative option as a comment here which retains the simplicity? I'd like to take a look

johnmccabe replied:
As a simple solution would adding a comment with the binary return type example?

johnmccabe replied:
Scratch that, wouldn't work... Unless we alter the `new` command to take arguments, `new --ascii`, `new --json`, `new --binary` etc. Hackiet option would be having those flags as suffixes to different template files

johnmccabe replied:
@alexellis wdyt? ^

andrew-s replied:
It's defining something I guess that's forcing type, I think the application should be more fluid because of the multi-use of the function. In theory, a custom class could be made and then the subsequent function could ask for the type based on that but, that's adding a layer of "custom code" on top. I think maybe leaving this open and not enforcing a type (as to, leave it to the application) would be a better option?

alexellis commented on template/php/template.yml:
```diff
@@ -0,0 +1,2 @@
+language: php
```
Do we want `php` or `php7`?

andrew-s replied:
So, the image will auto-roll up the minor versions of 7 - it's a question of, what do we do when there's php8?

We'll likely not support php5 as there's just no good reasoning for writing new code under that. Personally, I'd like to see us go php7 and version the templates - maybe there's a bigger architectural view there though? the language then sub-directories of supported versions? Not sure.

johnmccabe replied:
Agree with being explicit in the template name, `php7`, doesn't commit us to supporting anything now, but makes the introduction of future templates easier.

As for exposing versions via the tool (`--lang php --lang-version 7`) I think thats decoupled from where they get stored in the template repo, you'd ultimately want to have a per-version dir for the template and would just be abstracting away the choice.

andrew-s replied:
Are we all in agreement this is to be renamed php7 and to always be the latest stable php release?

johnmccabe replied:
+1 to php7 rename

johnmccabe replied:
+1 to php7 rename

Brunty replied:
Aye, php7 is the way to go

alexellis commented:
Andrew thanks for your work here.

Can you provide a help message for the user?

i.e. https://github.com/openfaas/templates/blob/master/template/go/template.yml#L3

alexellis commented:
@itscaro any comments for Andrew on this?

alexellis commented:
@martindekov could we get your input on this template as a PHP developer? Thanks

martindekov commented:
I think PHP is widely used and approve adding it to the official templates. PHP 7.2 will have active support for at least year+ so I don't worry about the version and we have python 3 and 2.7 which means we can have the same approach if PHP 8 comes out like I think @andrew-s pointed out. Also latest composer is `1.7.2` but we can update that later.

alexellis commented:
Ping @andrew-s - if you'd like this to move forward for a merge please can you address the feedback on the Dockerfile? Do you have an ETA?

John @johnmccabe is going to be doing a keynote talk at PHP Scotland in October. John, can you do some performance testing with this configuration using `hey` and share on Slack vs Go/Python3?

This template is using the classic watchdog so is forking per request. Do we know of a good HTTP server written in PHP which could be embedded as an of-watchdog version for performance comparison?

Alex

johnmccabe commented on template/php/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+
+# Function
+COPY index.php php-extension.sh composer.* ./
+RUN chmod u+x ./php-extension.sh && ./php-extension.sh
```
@andrew-s is executing the `php-extension.sh` here necessary, you're running again on line 36 below again.

johnmccabe commented:
Worth adding a README.md to the template with more info on setting up depdencies?

### php-extension.sh
- can install dependencies with `apk add`
- can use the helper utils `docker-php-ext-install `, `docker-php-ext-configure`, `docker-php-ext-enable ` etc
- there are a set of core extensions loaded by default in the base image
- point to the https://hub.docker.com/_/php/ instructions which we're exposing here

### Composer
- This one should be more obvious to users, but if adding a README for some of the stuff above I'd include a link to the composer docs as well.

andrew-s commented:
I'll look at getting the feedback implemented this week.

andrew-s commented:
> Worth adding a README.md to the template with more info on setting up depdencies?
> 
> ### php-extension.sh
> * can install dependencies with `apk add`
> * can use the helper utils `docker-php-ext-install`, `docker-php-ext-configure`, `docker-php-ext-enable` etc
> * there are a set of core extensions loaded by default in the base image
> * point to the https://hub.docker.com/_/php/ instructions which we're exposing here
> 
> ### Composer
> * This one should be more obvious to users, but if adding a README for some of the stuff above I'd include a link to the composer docs as well.

Yeah, I was looking at adding it into the official docs of how to use the template.

alexellis commented on template/php/template.yml:
```diff
@@ -0,0 +1,2 @@
+language: php
+fprocess: php index.php
```
In this file you can specify a welcome message explaining how to manage dependencies and which file to edit. See the others as examples.

johnmccabe replied:
I'd keep the message here brief and simple, and add more details into the `php-extension.sh`, something like:

```
welcome_message: |
  You have created a new function which uses PHP 7.2.
  Dependencies and extensions can be added using the composer.yml
  and php-extension.sh files.
```

alexellis commented on template/php/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+RUN [[ -f composer.lock || -f composer.json ]] && composer install --no-dev || \
+    echo "composer.lock does not exist, skip installing composer dependencies"
+
+# Cleanup
```
```
[0] > Pushing php [alexellis2/php:latest].
The push refers to repository [docker.io/alexellis2/php]
3bc9624fcc69: Pushed 
9622b5e05cc0: Pushed 
51918c3ecf8b: Pushing [============>                                      ]  5.234MB/20.42MB
b08f2129c755: Pushed 
ab2d2de3773c: Pushed 
76ea1f23a5b9: Pushing [==================================================>]  4.096kB
5508908f68c8: Pushing [==================================================>]  11.78kB
dc45c502d6dd: Pushing [>                                                  ]  67.07kB/4.453MB
fb86b2273725: Pushing [>                                                  ]  67.07kB/4.453MB
386e942947b2: Waiting 
d61378fdc4ce: Waiting 
d639c1323ea3: Waiting 
f1aefe9f02bb: Waiting 
861b714cd9d5: Waiting 
d73042a7a71a: Waiting 
75203553d20b: Waiting 
82c67d0a2ab7: Waiting 
51dcf0366b9d: Waiting 
ae1a20d1ae91: Waiting 
df64d3292fd6: Waiting 
```

^ A lot of these layers should collapse when this is finished.

johnmccabe commented on template/php/function/php-extension.sh:
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+echo "Installing PHP extensions"
+# Add your extensions in here
```
Add some text here explaining that you can use the helpers from the Docker library PHP image, and that its also possible to `apk` add stuff here as well where you have lib dependencies.

andrew-s commented on template/php/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+# Install Dependecies
+RUN [[ -f php-extension.sh ]] && sh ./php-extension.sh || \
+    echo "php-extension.sh does not exist, skip installing PHP extensions"
+RUN [[ -f composer.lock || -f composer.json ]] && composer install --no-dev || \
```
Yeah, I was thinking that when I wrote it, would be a little bit more verbose but, you're right

andrew-s commented on template/php/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+    echo "composer.lock does not exist, skip installing composer dependencies"
+
+# Cleanup
+RUN apk del git
```
As in,`--no-cache` during the install?

johnmccabe replied:
See - https://github.com/openfaas/templates/blob/master/template/go/Dockerfile#L7

andrew-s replied:
@johnmccabe I can add this in, that isn't a problem but, I'm struggling to find anything online that this does anything? I don't understand a `--no-cache` on a delete since it's a delete? Happy to learn more about how apk works though!

johnmccabe replied:
Likewise, its in use in the other templates, probably a cut and paste error? It actually ends up pulling down the cache index on delete !!
```
/ # apk del nginx -v --no-cache
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/community/x86_64/APKINDEX.tar.gz
...
```

johnmccabe commented on template/php/php-extension.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+
+echo "Skip installing PHP extensions for Core"
```
Since you're skipping this should just be removed (as mentioned in the Dockerfile comments).

johnmccabe commented on template/php/Dockerfile:
```diff
@@ -0,0 +1,52 @@
+COPY ./function ./function
+WORKDIR /home/app/function
+
+# Install Dependecies
```
Minor typo

alexellis commented:
@andrew-s that commit message isn't going to work. Please see the contribution guide on this topic, edit the message and do a force push into your branch.

andrew-s commented:
@alexellis Done, it was signed but my PGP key was missing from GitHub.

alexellis commented:
PGP keys are not required for contributions. This is explained in the contribution guide.

The problem was that it was not signed-off with your email and name. The second issue is the commit message is not worded correctly and is also truncated.

![bad-commit](https://user-images.githubusercontent.com/6358735/46157068-d62af580-c272-11e8-900b-e3769d501ff3.png)

Please read the commit message guidelines here -> https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md#commit-messages

alexellis commented:
This is what we need the commit messages to look like (which do follow the guidelines)

https://github.com/openfaas/templates/commit/e12758ae2717107a5837da3078fc3fc2a5decc4c

https://github.com/openfaas/templates/commit/8dea665fe025d3b887afa042d319559f55b5e7aa

Please edit and get back to us on this thread. I'm sure @johnmccabe or @ivanayov would be happy to answer questions on Slack.

johnmccabe commented:
@andrew-s heres a quick way to fix it, you can automatically add the signoff to your last 3 commits (note the `HEAD~3`) with a recent version of git and the following command:

```
git rebase --signoff HEAD~3
```

This is what it does:

![signoff-old-commits](https://user-images.githubusercontent.com/83862/46163037-f2825e80-c281-11e8-92f2-f9d8c2a3b077.gif)

johnmccabe commented:
I'm not going to be able to look at the changes until I get back from the theatre much later tonight I'm afraid, will do so before I head to bed tho.

alexellis commented:
I can see the commits as signed now, but the message style on `9007afe` hasn't changed.

alexellis commented:
Easiest thing to do is `git commit -s --amend` and then fix the last message.

andrew-s commented:
@alexellis Updated the commit message, apologies the issue was mainly time. Hope that works better.

johnmccabe commented:
I'm reviewing the updates now, cracking work on this PR.

johnmccabe commented on template/php7/Dockerfile:
```diff
@@ -0,0 +1,48 @@
+
+# Import function
+WORKDIR /home/app
+COPY index.php composer.* ./
```
@andrew-s there won't be a any `composer.*` files here surely, just copy `index.php` here?

andrew-s replied:
Yeah, missed that, updated and submitted.

johnmccabe commented on template/php7/Dockerfile:
```diff
@@ -0,0 +1,48 @@
+RUN [[ -f php-extension.sh ]] && sh ./php-extension.sh
+
+# Cleanup
+RUN apk del git && \
```
The ordering has changed here so `git` gets deleted before `composer install`, does Composer need git in any scenarios, I see it execing out to a local `git` binary [here](https://github.com/composer/composer/blob/master/src/Composer/Util/Git.php)? If it's worth keeping in then I'd just shift the cleanup to after the `composer install`, the extra `USER` switches aren't a big deal.

andrew-s replied:
Good spot, yeah, there's a few cases where you would need composer to have Git (e.g. private repo's, etc..)

johnmccabe commented on template/php7/readme.md:
````diff
@@ -0,0 +1,37 @@
+## Usage:
+
+```shell
+faas-cli new my-function-php7 --lang php7
````
Super picky thing, but I'd drop `-php7` from the function name just in case someone might misinterpret that as needed.

johnmccabe commented on template/php7/readme.md:
````diff
@@ -0,0 +1,37 @@
+faas-cli new my-function-php7 --lang php7
+```
+
+You will find in the newly created directories `my-function-in-php5` and `my-function-in-php7`:
````
These dirs are surely incorrect, you'd only have `my-function` (with the rename in the previous comment).

johnmccabe commented on template/php7/readme.md:
```diff
@@ -0,0 +1,37 @@
+If you need to install Phalcon for example, check out the following gist - you can use
+that file in your function into your `src/php-extension.sh` file;
+
+https://gist.github.com/andrew-s/07e152b7f981e8453da8d2d991d9aab4
```
I'd prefer this be checked in as `php-extension.sh-example` and linked here.

andrew-s replied:
Sure, I'm unsure how much this would be used but, I'd imagine the default extensions probably suffice for most people anyway. Do we agree on renaming?

andrew-s replied:
Are you suggesting to putting this into `php-extensions.sh` and renaming that file into `.sh-example`? or just the first?

johnmccabe replied:
Nope, I suggest putting in a file `template/php7/php-extensions.sh-example`, ie one in the repo, which can be linked to from the README.md in the same dir, and won't appear in the templated function that gets created on `new`, make sense?

Rationale is to keep everything in one place, much easier to update and manage.

johnmccabe commented:
Just a few final comments, but its looking 👍. Thanks for helping us get this over the line this week @andrew-s.

@alexellis I've got the test of-cloud fn ready for validation when this gets merged.

johnmccabe commented on template/php7/readme.md:
````diff
@@ -9,13 +9,13 @@ Templates are built using the latest minor version of the major release.
 ```
 
-You will find in the newly created directories `my-function-in-php5` and `my-function-in-php7`:
+You will find in the newly created directory `my-function-in-php`:
````
Directory is still wrong here I'm afraid, `-in-`. Would really prefer we drop the `-php` in the function name, when consuming a function one of the points is that the user shouldn't have to know or care what language its implemented in. Ie you could replace `my-function` with a Ruby, Python or whatever implementation and you don't have to change any code using that function.

andrew-s replied:
Makes sense, I'll make the change.

alexellis commented:
Derek lock: great to see this collaboration, for further discussion please join #templates on Slack
